File: A&DS/pcms9/task6.py
from sys import *
import threading

setrecursionlimit(10 ** 9)
threading.stack_size(2 ** 27)


def main():
    def dfs(node):
        for i in range(len(availableWays[node])):
            updNode = availableWays[node][i]
            if infoAboutNodes[updNode] == -1:
                dfs(updNode)
        for i in range(len(availableWays[node])):
            infoAboutNodes[node] = max(infoAboutNodes[node], (infoAboutNodes[availableWays[node][i]] + 1) % 2)
        if infoAboutNodes[node] == -1:
            infoAboutNodes[node] = 0  # if there aren`t any moves from node, first player always wins

    fin = open("game.in", "r")
    fout = open("game.out", "w")

    numOfNodes, lines, start = [int(x) for x in fin.readline().split()]
    availableWays = [[] for _ in range(numOfNodes)]
    infoAboutNodes = [-1 for _ in range(numOfNodes)]  # None - didn`t attend node, 1 - first player, 0 - second %2

    for i in range(lines):
        startNode, endNode = [int(x) for x in fin.readline().split()]
        availableWays[startNode - 1].append(endNode - 1)

    dfs(start-1)
    if infoAboutNodes[start-1] != 0:
        print("First player wins", file=fout)
    else:
        print("Second player wins", file=fout)


thread = threading.Thread(target=main)
thread.start()

File: A&DS/pcms12/task6.cpp
#include <iostream>
#include <vector>
#include <fstream>
using namespace std;

void dfs(int v, vector<vector<int>>& availableWays, vector<pair<int, int>>& d)
{
    for (auto u : availableWays[v])
    {
        dfs(u, availableWays, d);
    }
    for (auto u : availableWays[v])
    {
        d[v].second += d[u].first;
        d[v].first += max(d[u].first, d[u].second);
    }
}

int main()
{
    ifstream fin("selectw.in");
    ofstream fout("selectw.out");

    int numOfNodes, root;
    fin >> numOfNodes;
    vector<pair<int, int>> d;
    vector<vector<int>> availableWays(numOfNodes);

    for (int i = 0; i < numOfNodes; i++)
    {
        int node, value;
        fin >> node >> value;
        if (node != 0)
        {
            node--;
            availableWays[node].push_back(i);
        }
        else
        {
            root = i;
        }
        d.emplace_back(0, value);
    }

    dfs(root, availableWays, d);

    fout << max(d[root].first, d[root].second);
    return 0;
}
